### PR TITLE
fix(octez): implement de/serialize for bootstrap account

### PR DESCRIPTION
**Pull request by huancheng-trili**
# Context

Part of JSTZ-149.

[JSTZ-149](https://linear.app/tezos/issue/JSTZ-149/implement-protocol)

# Description

Replace `try_from` with de/serialize implementation for BootstrapAccount as done in #576.

# Manually testing the PR

* Unit test: added new test cases
```sh
$ cargo test --lib async::bootstrap::tests::deserialize_bootstrap_account
running 6 tests                                                                                                                                                       
test r#async::bootstrap::tests::deserialize_bootstrap_account_value_out_of_range ... ok                                              
test r#async::bootstrap::tests::deserialize_bootstrap_account ... ok              
test r#async::bootstrap::tests::deserialize_bootstrap_account_invalid_balance_type ... ok
test r#async::bootstrap::tests::deserialize_bootstrap_account_invalid_public_key ... ok
test r#async::bootstrap::tests::deserialize_bootstrap_account_incorrect_amount_of_elements ... ok
test r#async::bootstrap::tests::deserialize_bootstrap_accounts ... ok

test result: ok. 6 passed; 0 failed; 0 ignored; 0 measured; 54 filtered out; finished in 0.00s
$ cargo test --lib async::bootstrap::tests::serialize_bootstrap_account
running 2 tests                                                                                                                                                       
test r#async::bootstrap::tests::serialize_bootstrap_account ... ok                                                                                                    
test r#async::bootstrap::tests::serialize_bootstrap_accounts ... ok  

test result: ok. 2 passed; 0 failed; 0 ignored; 0 measured; 58 filtered out; finished in 0.00s
```

### CODE REVIEW

File: crates/octez/src/async/bootstrap.rs [hunk 1]
```diff
@@ -5,57 +5,88 @@ use serde::de::{Error, Visitor};
...
 
 #[derive(Clone, PartialEq, Eq, Debug)]
 pub struct BootstrapAccount {
     public_key: PublicKey,
     amount_mutez: u64,
 }
 
-impl TryFrom<&Value> for BootstrapAccount {
-    type Error = anyhow::Error;
+impl Serialize for BootstrapAccount {
```
Comment (zcabter) [inline crates/octez/src/async/bootstrap.rs:16]: Could you add a comment on how bootstrap accounts are represented in the params file?

File: crates/octez/src/async/bootstrap.rs [hunk 2]
```diff
@@ -5,57 +5,88 @@ use serde::de::{Error, Visitor};
...
+
+    fn visit_seq<A>(self, mut seq: A) -> Result<Self::Value, A::Error>
+    where
+        A: serde::de::SeqAccess<'de>,
+    {
+        if let Some(v) = seq.size_hint() {
+            if v != 2 {
+                return Err(A::Error::custom(
+                    "the input list should contain exactly two elements: address public key and account balance as string",
+                ));
```
Comment (zcabter) [inline crates/octez/src/async/bootstrap.rs:45]: `return` can be avoided by putting at the end `?`

File: crates/octez/src/async/bootstrap.rs [hunk 3]
```diff
@@ -5,57 +5,88 @@ use serde::de::{Error, Visitor};
...
-                "'amount' is not a valid string representing an unsigned integer"
-            ))?
-            .parse::<u64>();
-        if amount_mutez.is_err() {
-            return Err(anyhow::anyhow!(
-                "'amount' is not a valid string representing an unsigned integer"
+
+        if first_element.is_none() || second_element.is_none() {
+            return Err(A::Error::custom(
+                "the input list should contain exactly two elements: address public key and account balance as string",
```
Comment (zcabter) [inline crates/octez/src/async/bootstrap.rs:65]: nit: Slightly cleaner if the repeated error messages were moved into their own variable and reused

File: crates/octez/src/async/bootstrap.rs [hunk 4]
```diff
@@ -5,57 +5,88 @@ use serde::de::{Error, Visitor};
...
+        let mut first_element: Option<String> = None;
+        let mut second_element: Option<String> = None;
+        while let Some(current) = seq.next_element()? {
+            if first_element.is_none() {
+                first_element.replace(current);
+            } else if second_element.is_none() {
+                second_element.replace(current);
+            } else {
+                return Err(A::Error::custom(
+                    "the input list should only contain two elements",
```
Comment (zcabter) [inline crates/octez/src/async/bootstrap.rs:58]:
```suggestion
                    "the input list should only contain two elements but more found",
```

**Comment by zcabter** [comment]
Small comments, LGTM otherwise

File: crates/octez/src/async/bootstrap.rs (hunk 1 above)
Comment (huancheng-trili) [reply crates/octez/src/async/bootstrap.rs:16]: Done

File: crates/octez/src/async/bootstrap.rs (hunk 2 above)
Comment (huancheng-trili) [reply crates/octez/src/async/bootstrap.rs:49]: Updated

File: crates/octez/src/async/bootstrap.rs (hunk 4 above)
Comment (huancheng-trili) [reply crates/octez/src/async/bootstrap.rs]: Updated

**Review by zcabter — APPROVED** [review]

**Pull request merged**